### PR TITLE
Fix/sonar issues

### LINKED ISSUE #317: Items deleted when stacked in Building UIs
### Minecolonies version

minecolonies-universal-1.10.2-0.4.1870

Tested with
Forge 12.18.2.2179 (latest)
Forge 12.18.2.2123 (One I use)
Forge 12.18.2.2099 (Recommended)

### Expected behavior

- Adding items to a stack in a Building Inventory should increase stack size.

### Actual behaviour

- Stack size appears to increase, but interacting with stack sets in back to original value. Added items are lost.

### Steps to reproduce the problem

1. Put 16 dirt in Builder's Hut Inventory.
2. Add 10 more to that stack.
3. Stack displays 26.
4. Click stack to pick up.
5. You are the proud owner of 16 dirt.

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #317 and will fix up a lot of sonar issues + miner and builder issues

@Minecolonies/maintainers 
Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/blocks/BlockSolidSubstitution.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+    /**
+     * Used to determine ambient occlusion and culling when rebuilding chunks for render.
+     *
+     * @return true
+     */
+    //todo: remove once we no longer need to support this
+    @SuppressWarnings("deprecation")
+    @Override
+    public boolean isOpaqueCube(final IBlockState state)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/blocks/BlockSolidSubstitution.java:67]: MINOR Don't override a deprecated method or explicitly mark it as "@Deprecated". [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/Level.java [hunk 2]
```diff
@@ -46,47 +95,117 @@ private Level()
...
+        for(final Point2D.Double pos: nodeCenterList)
         {
-            ladderNode.setDirectionNegZ(Node.NodeStatus.LADDER);
-            cobbleNode.setDirectionPosZ(Node.NodeStatus.LADDER);
+            if(nodes.containsKey(pos))
+            {
+                continue;
+            }
+            final Node tempNodeToAdd = new Node(pos.getX(), pos.getY(), new Point2D.Double(tempNode.getX(), tempNode.getZ()));
+            int randNumber = rand.nextInt(RANDOM_TYPES);
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/Level.java]: MINOR Local variable 'randNumber' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/Level.java [hunk 3]
```diff
@@ -46,47 +95,117 @@ private Level()
...
+            tempNodeToAdd.setStyle(randNumber <= 1 ? Node.NodeType.TUNNEL : (randNumber == 2 ? Node.NodeType.BEND : Node.NodeType.CROSSROAD));
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/Level.java]: MAJOR Unnecessary use of fully qualified name 'Node.NodeType.TUNNEL' due to existing static import 'com.minecolonies.coremod.entity.ai.citizen.miner.Node.NodeType.TUNNEL' [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3AUnnecessaryFullyQualifiedName)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java [hunk 4]
```diff
@@ -143,101 +371,179 @@ check if we have not found a block (when block == null
...
     }
 
+
     /**
      * Works on clearing the area of unneeded blocks.
      *
      * @return the next step once done.
      */
-    private boolean clearStep(@NotNull final Structure.SchematicBlock currentBlock)
+    private boolean clearStep(@NotNull final Structure.StructureBlock currentBlock)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java:516]: MAJOR The Cyclomatic Complexity of this method "clearStep" is 13 which is greater than 10 authorized. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java [hunk 5]
```diff
@@ -143,101 +371,179 @@ check if we have not found a block (when block == null
...
         {
-            @NotNull final BlockPos positionInDirection = getPositionInDirection(direction, distance);
-            if (EntityUtils.checkForFreeSpace(world, positionInDirection))
+            while (((JobBuilder) job).getStructure().findNextBlock())
             {
-                return positionInDirection;
+                @Nullable final Block block = ((JobBuilder) job).getStructure().getBlock();
+                @NotNull final IBlockState blockState = ((JobBuilder) job).getStructure().getBlockState();
+
+                if (((JobBuilder) job).getStructure().doesStructureBlockEqualWorldBlock() ||
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java]: MINOR '||' should be on a new line. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AOperator_Wrap)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java [hunk 6]
```diff
@@ -66,60 +109,244 @@ protected AbstractEntityAIStructure(@NotNull final J job)
...
+            worker.addExperience(XP_EACH_BUILDING);
+            workFrom = null;
+        }
+
+        currentStructure = null;
+
+        return AIState.IDLE;
+    }
+
+    private Boolean decorationStep(final Structure.StructureBlock structureBlock)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java:232]: MAJOR The Cyclomatic Complexity of this method "decorationStep" is 13 which is greater than 10 authorized. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java [hunk 7]
```diff
@@ -66,60 +109,244 @@ protected AbstractEntityAIStructure(@NotNull final J job)
...
+                if (wo == null)
+                {
+                    Log.getLogger().error(String.format("Builder (%d:%d) ERROR - Finished, but missing work order(%d)",
+                            worker.getColony().getID(),
+                            worker.getCitizenData().getId(),
+                            ((JobBuilder) job).getWorkOrderId()));
+                }
+                else
+                {
+                    if (!(wo instanceof WorkOrderBuildDecoration))
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java]: MAJOR Avoid if (x != y) ..; else ..; [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3AConfusingTernary)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java [hunk 8]
```diff
@@ -143,101 +371,179 @@ check if we have not found a block (when block == null
...
+    public void loadStructure(@NotNull final String name, int rotateTimes, BlockPos position)
     {
-        if (offset > MAX_ADDITIONAL_RANGE_TO_BUILD)
+        if(job instanceof AbstractJobStructure)
         {
-            return currentStructure.getCurrentBlockPosition();
+            rotation = rotateTimes;
+            try
+            {
+                 StructureWrapper wrapper = new StructureWrapper(world, name);
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java]: MINOR Local variable 'wrapper' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java [hunk 9]
```diff
@@ -66,60 +109,244 @@ protected AbstractEntityAIStructure(@NotNull final J job)
...
           /**
            * Finalize the building and give back control to the ai.
-           * todo: implement
            */
-          new AITarget(AIState.COMPLETE_BUILD, () -> AIState.IDLE)
+          new AITarget(COMPLETE_BUILD, this::completeBuild)
         );
     }
 
+    private AIState completeBuild()
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java:142]: MAJOR The Cyclomatic Complexity of this method "completeBuild" is 12 which is greater than 10 authorized. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java [hunk 10]
```diff
@@ -66,60 +109,244 @@ protected AbstractEntityAIStructure(@NotNull final J job)
...
+                {
+                    ((BuildingBuilder) workerBuilding).resetNeededResources();
+                }
+                resetTask();
+            }
+            else if(job instanceof JobMiner)
+            {
+                BuildingMiner minerBuilding = (BuildingMiner) getOwnBuilding();
+                //If shaft isn't cleared we're in shaft clearing mode.
+                if(!minerBuilding.clearedShaft)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java]: MAJOR Avoid if (x != y) ..; else ..; [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3AConfusingTernary)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java [hunk 11]
```diff
@@ -316,4 +694,223 @@ private AIState startBuilding()
...
+     * @param blockToMine block which should be mined or placed.
+     * @return the save position.
+     */
+    private BlockPos getNodeMiningPosition(BlockPos blockToMine)
+    {
+        if(((BuildingMiner) getOwnBuilding()).getCurrentLevel() == null)
+        {
+            return blockToMine;
+        }
+        Point2D pos = ((BuildingMiner) getOwnBuilding()).getCurrentLevel().getRandomNode().getParent();
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java]: MINOR Local variable 'pos' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java [hunk 12]
```diff
@@ -316,4 +694,223 @@ private AIState startBuilding()
...
+    /**
+     * Defines blocks that can be built for free.
+     *
+     * @param block    The block to check if it is free.
+     * @param metadata The metadata of the block.
+     * @return true or false.
+     */
+    public static boolean isBlockFree(@Nullable final Block block, final int metadata)
+    {
+        return block == null
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java:723]: MAJOR Reduce the number of conditional operators (8) used in the expression (maximum allowed 4). [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1067)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java [hunk 13]
```diff
@@ -66,60 +109,244 @@ protected AbstractEntityAIStructure(@NotNull final J job)
...
+            {
+                return false;
+            }
+
+            placeBlockAt(block, blockState, structureBlock.blockPosition);
+        }
+        return true;
+    }
+
+    private Boolean structureStep(final Structure.StructureBlock structureBlock)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java:276]: MAJOR Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)
MAJOR The Cyclomatic Complexity of this method "structureStep" is 19 which is greater than 10 authorized. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java [hunk 14]
```diff
@@ -316,4 +694,223 @@ private AIState startBuilding()
...
+
+            if (!placeBlock(coords, block, blockState))
+            {
+                Log.getLogger().error(String.format("Block place failure %s at %s", block.getUnlocalizedName(), coords));
+            }
+            worker.swingArm(worker.getActiveHand());
+        }
+    }
+
+    private boolean placeBlock(@NotNull final BlockPos pos, final Block block, @NotNull final IBlockState blockState)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java:756]: MAJOR Reduce the number of returns of this method 7, down to the maximum allowed 5. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)
MAJOR The Cyclomatic Complexity of this method "placeBlock" is 31 which is greater than 10 authorized. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java [hunk 15]
```diff
@@ -28,7 +53,14 @@
  */
 public abstract class AbstractEntityAIStructure<J extends AbstractJob> extends AbstractEntityAIInteract<J>
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java:54]: MAJOR Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 43 to the maximum authorized 30 or less. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1200)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java [hunk 16]
```diff
@@ -143,101 +371,179 @@ check if we have not found a block (when block == null
...
-        //get length or width either is larger.
-        final int length = currentStructure.getLength();
-        final int width = currentStructure.getWidth();
-        final int distance = Math.max(width, length) + MIN_ADDITIONAL_RANGE_TO_BUILD + offset;
-        @NotNull final EnumFacing[] directions = {EnumFacing.EAST, EnumFacing.WEST, EnumFacing.NORTH, EnumFacing.SOUTH};
+    }
 
-        //then get a solid place with two air spaces above it in any direction.
-        for (final EnumFacing direction : directions)
+    private void requestMaterials()
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java:474]: MAJOR The Cyclomatic Complexity of this method "requestMaterials" is 15 which is greater than 10 authorized. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: src/main/java/com/minecolonies/structures/helpers/Structure.java [hunk 17]
```diff
@@ -329,14 +329,14 @@ private void renderGhost(final World world, final ModelHolder holder, final Enti
                     this.mc.getTextureManager().bindTexture(TextureMap.LOCATION_BLOCKS_TEXTURE);
                     ForgeHooksClient.setRenderLayer(layer);
                     this.renderGhostBlock(world, holder, player, layer, existingModel, partialTicks);
-                    holder.rendered = true;
+                    holder.setRendered(true);;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/structures/helpers/Structure.java:332]: MINOR Remove this empty statement. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AEmptyStatementUsageCheck)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java [hunk 18]
```diff
@@ -842,539 +601,97 @@ private boolean secureBlock(@NotNull final BlockPos curBlock, @NotNull final Blo
...
+    /**
+     * Initiates structure loading.
+     * @param mineNode the node to load it for.
+     * @param direction the direction it faces.
+     */
+    private void initStructure(final Node mineNode, final int rotateTimes, BlockPos structurePos)
     {
-        //Preload structures
-        if (job.getStructure() == null)
+        if(mineNode != null)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java]: MAJOR Avoid if (x != y) ..; else ..; [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3AConfusingTernary)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java [hunk 19]
```diff
@@ -842,539 +601,97 @@ private boolean secureBlock(@NotNull final BlockPos curBlock, @NotNull final Blo
...
-                    @NotNull final Node newnode = createNewNodeInDirectionFromNode(currentNode, dir);
-                    currentLevel.addNode(newnode);
-                    return newnode;
-                }
-            }
+            return CLEAR_STEP;
         }
 
-        return null;
+        return AIState.MINER_MINING_NODE;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java]: MAJOR Unnecessary use of fully qualified name 'AIState.MINER_MINING_NODE' due to existing static import 'com.minecolonies.coremod.entity.ai.util.AIState.*' [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3AUnnecessaryFullyQualifiedName)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java [hunk 20]
```diff
@@ -607,216 +510,72 @@ private BlockPos getNextBlockInShaftToMine()
...
-        int zOffSet = getZDistance(foundDirection) / 2;
-        if (xOffSet > 0)
-        {
-            xOffSet += 1;
-        }
-        else
+        //normal facing +x
+        int rotation = 0;
+
+        int vectorX = workingNode.getX() < workingNode.getParent().getX() ? -1 : (workingNode.getX() > workingNode.getParent().getX() ? 1 : 0);
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java]: MINOR Local variable 'vectorX' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java [hunk 21]
```diff
@@ -607,216 +510,72 @@ private BlockPos getNextBlockInShaftToMine()
...
+        int vectorZ = workingNode.getZ() < workingNode.getParent().getY() ? -1 : (workingNode.getZ() > workingNode.getParent().getY() ? 1 : 0);
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java]: MINOR Local variable 'vectorZ' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/coremod/util/BlockUtils.java [hunk 22]
```diff
@@ -86,6 +87,27 @@ public static boolean shouldNeverBeMessedWith(final Block block)
...
+    public static int getRotationFromFacing(final EnumFacing facing)
+    {
+        switch (facing)
+        {
+            case SOUTH:
+                return 2;
+            case EAST:
+                return 1;
+            case WEST:
+                return 3;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/util/BlockUtils.java:104]: MINOR Assign this magic number 3 to a well-named constant, and use the constant instead. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/coremod/client/gui/WindowHutMiner.java [hunk 23]
```diff
@@ -21,16 +21,14 @@
...
-    private static final String BUTTON_NEXTPAGE           = "nextPage";
     private static final String BUTTON_CURRENTLEVEL       = "changeToLevel";
     private static final String VIEW_PAGES                = "pages";
     private static final String HUT_MINER_RESOURCE_SUFFIX = ":gui/windowHutMiner.xml";
-    private Button             buttonPrevPage;
-    private Button             buttonNextPage;
-    private int[]              levels;
-    private ScrollingList      levelList;
-    private BuildingMiner.View miner;
+    private       Button             buttonPrevPage;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowHutMiner.java]: MAJOR Remove this unused "buttonPrevPage" private field. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1068)

File: src/main/java/com/minecolonies/coremod/entity/ai/util/Structure.java [hunk 24]
```diff
@@ -80,14 +110,47 @@ public SchematicBlock(final Block block, final BlockPos blockPosition, final IBl
             this.worldBlock = worldBlock;
             this.worldMetadata = worldMetadata;
         }
+
+        /**
+         * Checks if the structureBlock equals the worldBlock.
+         * @return true if so.
+         */
+        public boolean doesStructureBlockEqualWorldBlock()
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/util/Structure.java:118]: MAJOR The Cyclomatic Complexity of this method "doesStructureBlockEqualWorldBlock" is 13 which is greater than 10 authorized. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: src/main/java/com/minecolonies/coremod/entity/ai/util/Structure.java [hunk 25]
```diff
@@ -51,14 +56,39 @@ public StructureException(final String message)
         }
     }
 
-    public static final class SchematicBlock
+    /**
+     * Class used to describe a certain structure block of the structure.
+     */
+    public static final class StructureBlock
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/util/Structure.java:62]: MAJOR Reduce this class from 85 to the maximum allowed 50 or externalize it in a public class. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS2972)

File: src/main/java/com/minecolonies/coremod/util/RecipeHandler.java [hunk 26]
```diff
@@ -43,6 +43,8 @@ public static void init(final boolean enableInDevelopmentFeatures, final boolean
         GameRegistry.addRecipe(new ItemStack(ModItems.scanTool, 1), "  I", " S ", "S  ", 'I', Items.IRON_INGOT, 'S', Items.STICK);
         GameRegistry.addRecipe(new ItemStack(ModItems.buildTool, 1), "  C", " S ", "S  ", 'C', Blocks.COBBLESTONE, 'S', Items.STICK);
         GameRegistry.addRecipe(new ItemStack(ModBlocks.blockSubstitution, 16), "XXX", "X#X", "XXX", 'X', Blocks.PLANKS, '#', ModItems.scanTool);
+        GameRegistry.addRecipe(new ItemStack(ModBlocks.blockSolidSubstitution, 16), "XXX", "X#X", "XXX", 'X', Blocks.LOG, '#', ModItems.scanTool);
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/util/RecipeHandler.java:46]: MINOR Assign this magic number 16 to a well-named constant, and use the constant instead. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java [hunk 27]
```diff
@@ -204,754 +176,13 @@ private void initiate()
...
-        if (workerBuilding instanceof BuildingBuilder)
-        {
-            ((BuildingBuilder) workerBuilding).resetNeededResources();
-        }
-        resetTask();
-        worker.addExperience(XP_EACH_BUILDING);
-        workFrom = null;
-
-        return AIState.IDLE;
+        return AIState.START_BUILDING;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java]: MAJOR Unnecessary use of fully qualified name 'AIState.START_BUILDING' due to existing static import 'com.minecolonies.coremod.entity.ai.util.AIState.*' [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3AUnnecessaryFullyQualifiedName)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java [hunk 28]
```diff
@@ -93,21 +56,30 @@ public EntityAIStructureBuilder(@NotNull final JobBuilder job)
...
         );
         worker.setSkillModifier(INTELLIGENCE_MULTIPLIER * worker.getCitizenData().getIntelligence()
                                   + STRENGTH_MULTIPLIER * worker.getCitizenData().getStrength());
         worker.setCanPickUpLoot(true);
     }
 
+    @Override
+    public Block getSolidSubstitution(BlockPos location)
+    {
+        IBlockState filler = world.getBiome(location).fillerBlock;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java]: MINOR Local variable 'filler' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 40 issues
* MAJOR 21 major
* MINOR 19 minor

Watch the comments in this conversation to review them.

#### 3 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [Structure.java#L302](https://github.com/Minecolonies/minecolonies/blob/5d4ec23d24b2ab45d80dd911fe4a2e3c6e71ad6c/src/main/java/com/minecolonies/coremod/entity/ai/util/Structure.java#L302): The Cyclomatic Complexity of this method "advanceBlock" is 12 which is greater than 10 authorized. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. MINOR [WindowHutBuilder.java#L4](https://github.com/Minecolonies/minecolonies/blob/5d4ec23d24b2ab45d80dd911fe4a2e3c6e71ad6c/src/main/java/com/minecolonies/coremod/client/gui/WindowHutBuilder.java#L4): Remove this unused import 'com.minecolonies.blockout.controls.Button'. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [WindowHutDeliveryman.java#L4](https://github.com/Minecolonies/minecolonies/blob/5d4ec23d24b2ab45d80dd911fe4a2e3c6e71ad6c/src/main/java/com/minecolonies/coremod/client/gui/WindowHutDeliveryman.java#L4): Remove this unused import 'com.minecolonies.blockout.views.SwitchView'. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 19 issues
* MAJOR 13 major
* MINOR 6 minor

Watch the comments in this conversation to review them.

#### 3 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [Structure.java#L302](https://github.com/Minecolonies/minecolonies/blob/f774d5c1508ec0eeb4543f69a7e604acff52765b/src/main/java/com/minecolonies/coremod/entity/ai/util/Structure.java#L302): The Cyclomatic Complexity of this method "advanceBlock" is 12 which is greater than 10 authorized. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. MINOR [WindowHutBuilder.java#L4](https://github.com/Minecolonies/minecolonies/blob/f774d5c1508ec0eeb4543f69a7e604acff52765b/src/main/java/com/minecolonies/coremod/client/gui/WindowHutBuilder.java#L4): Remove this unused import 'com.minecolonies.blockout.controls.Button'. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [WindowHutDeliveryman.java#L4](https://github.com/Minecolonies/minecolonies/blob/f774d5c1508ec0eeb4543f69a7e604acff52765b/src/main/java/com/minecolonies/coremod/client/gui/WindowHutDeliveryman.java#L4): Remove this unused import 'com.minecolonies.blockout.views.SwitchView'. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/minecolonies/structures/helpers/Structure.java (hunk 17 above)
Comment (marvin-bitterlich) [reply src/main/java/com/minecolonies/structures/helpers/Structure.java:332]: @Raycoms can you so this? 😛

**Comment by marvin-bitterlich** [comment]
LGTM!

[Approved with PullApprove](https://pullapprove.com/Minecolonies/minecolonies/pull-request/334/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)

File: src/main/java/com/minecolonies/structures/helpers/Structure.java (hunk 17 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/structures/helpers/Structure.java:332]: so?

**Pull request merged**